tiles/src/main/java/com/protomaps/basemap/text/TextEngine.java

```diff
@@ -0,0 +1,190 @@
+      String glyphKey = FontRegistry.getGlyphKey(index, xOffset + deltaXOffset, yOffset, xAdvance + deltaXAdvance, yAdvance);
+      if (encoding.get(glyphKey) != null) {
+        return encoding.get(glyphKey);
+      }
+    }
+    // TODO: handle error
+    System.out.println("Could not find a matching glyph for index = " + index +
+      ", xOffset = " + xOffset +
+      ", yOffset = " + yOffset +
+      ", xAdvance = " + xAdvance +
+      ", yAdvance = " + yAdvance + ". Aborting...");
+    System.exit(1);
+    return 0; // did not find any matching codepoint
+  }
+
+  public static String encode(String text, Font font, HashMap<String, Integer> encoding) {
+    String result = "";
+
+    FontRenderContext frc = new FontRenderContext(null, true, true);
+    char[] charArray = text.toCharArray();
+    GlyphVector glyphVector = font.layoutGlyphVector(frc, charArray, 0, charArray.length, 0);
+
+    float sumXAdvances = 0;
+
+    for (int i = 0; i < glyphVector.getNumGlyphs(); i++) {
+      GlyphMetrics glyphMetrics = glyphVector.getGlyphMetrics(i);
+      int glyphCode = glyphVector.getGlyphCode(i);
+
+      double xAdvance = glyphMetrics.getAdvanceX();
+      double xPosition = glyphVector.getGlyphPosition(i).getX();
+      double xOffset = xPosition - sumXAdvances;
+
+      int xAdvanceML = (int) Math.floor(1000.0 * xAdvance / 64.0);
+      int xOffsetML = (int) Math.floor(1000.0 * xOffset / 64.0);
+
+      int yAdvanceML = 0;
+      int yOffsetML = 0;
+
+      int codepoint = codepointFromGlyph(encoding, glyphCode, xOffsetML, yOffsetML, xAdvanceML, yAdvanceML);
+
+      sumXAdvances += xAdvance;
+
+      result += new StringBuilder().appendCodePoint(codepoint).toString();
+    }
+    return result;
+  }
+
+  public static List<String> segment(String text, List<String> scripts) {
+    List<String> segments = new ArrayList<>();
+
+    if (text == null || text.isEmpty()) {
+      return segments;
+    }
+
+    String inner = "";
+    for (String script : scripts) {
+      inner += "\\p{In" + script + "}";
+    }
+    String regex = "[" + inner + "]+|[^" + inner + "]+";
+
+    Pattern pattern = Pattern.compile(regex);
+    Matcher matcher = pattern.matcher(text);
+
+    while (matcher.find()) {
+      segments.add(matcher.group());
+    }
+
+    return segments;
+  }
+
+  public static String encodeRegisteredScripts(String text, FontRegistry fontRegistry) {
+    if (text == null || text.isEmpty()) {
+      return "";
+    }
+    List<String> segments = segment(text, fontRegistry.getScripts());
+    String encodedText = "";
+    for (String segment : segments) {
+      String script = Script.getScript(segment);
+      if (fontRegistry.getScripts().contains(script)) {
+        encodedText += TextEngine.encode(segment, fontRegistry.getFont(script), fontRegistry.getEncoding(script));
+      }
+      else {
+        encodedText += segment;
+      }
+    }
+    return encodedText;
+  }
+
+  public static void main(String[] args) {
```

Comment:
feels like we can move this `main` into a small unit or integration test instead of having multiple `mains` in the program?

Reply:
Moved to unit tests.